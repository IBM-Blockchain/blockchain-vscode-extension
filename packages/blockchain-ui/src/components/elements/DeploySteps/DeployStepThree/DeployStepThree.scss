html{
    overflow: overlay !important;
    
    #root {
        .deploy-page-container {
            #customize{
                p {
                    margin-bottom: 8px;
                }
            }
        }
    }
}