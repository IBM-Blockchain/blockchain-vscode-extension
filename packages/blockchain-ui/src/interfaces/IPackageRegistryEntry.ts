interface IPackageRegistryEntry {
    name: string;
    path: string;
    version?: string;
    sizeKB: number;
}

export default IPackageRegistryEntry;
