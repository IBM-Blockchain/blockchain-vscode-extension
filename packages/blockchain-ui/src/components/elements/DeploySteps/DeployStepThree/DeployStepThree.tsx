import React, { Component } from 'react';
import { Link, UnorderedList, ListItem, Accordion, AccordionItem, Toggle, MultiSelect, DataTable, TableContainer, Table, TableHead, TableRow, TableHeader, TableBody, TableCell } from 'carbon-components-react';
import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';

interface IProps {
    selectedPackage: IPackageRegistryEntry;
    channelName: string;
    commitSmartContract: undefined | boolean;
    selectedPeers: string[];
    discoveredPeers: string[];
    onCommitChange: (value: boolean) => void;
    onPeerChange: (peers: string[]) => void;
}

interface StepThreeState {
    showCommitListItem: boolean;
    selectedPeers: string[];
}

class DeployStepThree extends Component<IProps, StepThreeState> {
    constructor(props: Readonly<IProps>) {
        super(props);
        this.state = {
            showCommitListItem: this.props.commitSmartContract !== undefined ? this.props.commitSmartContract : true,
            selectedPeers: this.props.selectedPeers
        };

        this.toggleCommit = this.toggleCommit.bind(this);
        this.changePeers = this.changePeers.bind(this);
        this.formatDiscoveredPeers = this.formatDiscoveredPeers.bind(this);

    }

    toggleCommit(checked: boolean, _id: string, _event: any): void {
        this.setState({ showCommitListItem: checked });
        this.props.onCommitChange(checked);
    }

    changePeers(event: { selectedItems: { id: string; label: string; }[] }): void {
        const peers: string[] = event.selectedItems.map((peerObject: { id: string, label: string }) => {
            return peerObject.id;
        });
        this.props.onPeerChange(peers);
    }

    formatDiscoveredPeers(peers: string[]): { id: string, label: string }[] {
        return peers.map((_peer) => {
            return { id: _peer, label: _peer };
        });
    }

    render(): JSX.Element {
        const headerData: { header: string, key: string }[] = [
            {
                header: 'Organization',
                key: 'organization',
            },
            {
                header: 'Approval status',
                key: 'status',
            }
        ];
        const rowData: { id: string, organization: string, status: string }[] = [
            {
                id: 'a',
                organization: 'Hardcoded',
                status: 'Value'
            },
            {
                id: 'b',
                organization: 'Needs to be',
                status: 'Changed'
            }
        ];

        let commitListItem: JSX.Element = <></>;

        if (this.state.showCommitListItem) {
            commitListItem = <ListItem>Commit the definition to `{this.props.channelName}`</ListItem>;
        }

        let packageName: string;
        if (this.props.selectedPackage.version) {
            packageName = `${this.props.selectedPackage.name}@${this.props.selectedPackage.version}`;
        } else {
            packageName = `${this.props.selectedPackage.name}`;
        }

        const discoveredPeers: { id: string, label: string }[] = this.formatDiscoveredPeers(this.props.discoveredPeers);
        const selectedPeers: { id: string, label: string }[] = this.formatDiscoveredPeers(this.state.selectedPeers);

        return (
            <>
                <div className='bx--row margin-bottom-06'>
                    <div className='bx--col-lg-10'>
                        <span>When you select 'Deploy', the following actions will automatically occur:</span>

                        <UnorderedList className='padding-left-07 padding-top-05'>
                            <ListItem>
                                Install smart contract package `{packageName}` on all peers
                            </ListItem>
                            <ListItem>
                                Approve the same smart contract definition for each organization
                            </ListItem>
                            {commitListItem}

                        </UnorderedList>
                    </div>
                </div>
                <div className='bx--row margin-bottom-07'>
                    <div className='bx--col-lg-10'>
                        <Accordion id='advancedAccordion'>
                            <AccordionItem id='customize' title={'Customize commit (advanced)'}>
                                <div className='bx--row margin-top-06'>
                                    <div className='bx--col'>
                                        <p>Perform commit</p>
                                        <Toggle
                                            defaultToggled={this.state.showCommitListItem}
                                            id='commitToggle'
                                            labelA='Off'
                                            labelB='On'
                                            onToggle={this.toggleCommit}
                                        />
                                    </div>
                                    <div className='bx--col'>
                                        <p>Additional peers to endorse commit transactions</p>

                                        <MultiSelect
                                            id='peer-select'
                                            initialSelectedItems={selectedPeers}
                                            items={discoveredPeers}
                                            label='Select peers'
                                            onChange={this.changePeers}
                                        />

                                    </div>
                                </div>
                                <div className='bx--row margin-top-07'>
                                    <div className='bx--col'>
                                        <DataTable
                                            rows={rowData}
                                            headers={headerData}
                                            render={({ rows, headers, getHeaderProps }) => (
                                                <TableContainer>
                                                    <Table size='short'>
                                                        <TableHead>
                                                            <TableRow>
                                                                {headers.map((header) => (
                                                                    <TableHeader {...getHeaderProps({ header })}>
                                                                        {header.header}
                                                                    </TableHeader>
                                                                ))}
                                                            </TableRow>
                                                        </TableHead>
                                                        <TableBody>
                                                            {rows.map((row) => (
                                                                <TableRow key={row.id}>
                                                                    {row.cells.map((cell) => (
                                                                        <TableCell key={cell.id}>{cell.value}</TableCell>
                                                                    ))}
                                                                </TableRow>
                                                            ))}
                                                        </TableBody>
                                                    </Table>
                                                </TableContainer>)}
                                        />
                                    </div>
                                </div>
                                <div className='bx--row margin-top-07'>
                                    <div className='bx--col'>
                                        <p>For an explanation of advanced scenarios, see the <Link href='#'>documentation</Link>.</p>
                                    </div>
                                </div>
                            </AccordionItem>
                        </Accordion>

                    </div>
                </div>
            </>
        );
    }
}

export default DeployStepThree;
