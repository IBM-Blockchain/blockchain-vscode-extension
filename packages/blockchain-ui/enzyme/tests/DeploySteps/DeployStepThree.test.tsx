import React from 'react';
import renderer from 'react-test-renderer';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import DeployStepThree from '../../../src/components/elements/DeploySteps/DeployStepThree/DeployStepThree';
import IPackageRegistryEntry from '../../../src/interfaces/IPackageRegistryEntry';
import { ReactWrapper, mount } from 'enzyme';

chai.should();
chai.use(sinonChai);

describe('DeployStepThree component', () => {
    let mySandBox: sinon.SinonSandbox;

    const packageOne: IPackageRegistryEntry = {name: 'mycontract', version: '0.0.1', path: '/package/one', sizeKB: 9000};

    let commitChangeStub: sinon.SinonStub;
    beforeEach(async () => {
        mySandBox = sinon.createSandbox();
        commitChangeStub = mySandBox.stub();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    describe('render', () => {
        it('should render the expected snapshot', () => {
            const component: any = renderer
                .create(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />)
                .toJSON();
            expect(component).toMatchSnapshot();
        });

        it('should show commit list item if toggled on', async () => {
            const component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />);

            component.setState({showCommitListItem: true});

            const doesCommitListItemExist: any = component.text().includes('Commit the definition to `mychannel`');
            doesCommitListItemExist.should.equal(true);
        });

        it('should not show commit list item if toggled off', async () => {
            const component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />);

            component.setState({showCommitListItem: false});

            const doesCommitListItemExist: any = component.text().includes('Commit the definition to `mychannel`');
            doesCommitListItemExist.should.equal(false);
        });
    });

    describe('toggleCommit', () => {

        it('toggle should be set to true if no prop passed', async () => {
            const component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />);
            const instance: DeployStepThree = component.instance() as DeployStepThree;
            instance.state.showCommitListItem.should.equal(true);
        });

        it('toggle should be set to prop value if undefined', async () => {
            let component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={false} onCommitChange={commitChangeStub} />);
            let instance: DeployStepThree = component.instance() as DeployStepThree;
            instance.state.showCommitListItem.should.equal(false);

            component = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={true} onCommitChange={commitChangeStub} />);
            instance = component.instance() as DeployStepThree;
            instance.state.showCommitListItem.should.equal(true);
        });

        it('should set state correctly when toggled off and on again', async () => {

            const component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />);
            const instance: DeployStepThree = component.instance() as DeployStepThree;

            const setState: sinon.SinonStub = mySandBox.stub(instance, 'setState').resolves();

            instance.toggleCommit(false, 'some_id', {some: 'event'});

            setState.should.have.been.calledWithExactly({showCommitListItem: false});
            commitChangeStub.should.have.been.calledWith(false);

            instance.toggleCommit(true, 'other_id', {some: 'otherEvent'});

            setState.should.have.been.calledWithExactly({showCommitListItem: true});
            commitChangeStub.should.have.been.calledWith(true);
        });
    });

    describe('changePeers', () => {
        it('should set state when toggled', async () => {

            const component: ReactWrapper<DeployStepThree> = mount(<DeployStepThree selectedPackage={packageOne} channelName='mychannel' commitSmartContract={undefined} onCommitChange={commitChangeStub} />);
            const instance: DeployStepThree = component.instance() as DeployStepThree;

            // TODO: Update when code is implemented.

            instance.changePeers({selectedItems: [{id: 'hello', label: 'world'}]});

        });
    });

});
