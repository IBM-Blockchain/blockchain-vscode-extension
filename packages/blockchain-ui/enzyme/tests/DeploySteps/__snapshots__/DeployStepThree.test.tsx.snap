// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployStepThree component render should render the expected snapshot 1`] = `
Array [
  <div
    className="bx--row margin-bottom-06"
  >
    <div
      className="bx--col-lg-10"
    >
      <span>
        When you select 'Deploy', the following actions will automatically occur:
      </span>
      <ul
        className="bx--list--unordered padding-left-07 padding-top-05"
      >
        <li
          className="bx--list__item"
        >
          Install smart contract package \`
          mycontract@0.0.1
          \` on all peers
        </li>
        <li
          className="bx--list__item"
        >
          Approve the same smart contract definition for each organization
        </li>
        <li
          className="bx--list__item"
        >
          Commit the definition to \`
          mychannel
          \`
        </li>
      </ul>
    </div>
  </div>,
  <div
    className="bx--row margin-bottom-07"
  >
    <div
      className="bx--col-lg-10"
    >
      <ul
        className="bx--accordion bx--accordion--end"
        id="advancedAccordion"
      >
        <li
          className="bx--accordion__item"
          id="customize"
          onAnimationEnd={[Function]}
        >
          <button
            aria-expanded={false}
            className="bx--accordion__heading"
            onClick={[Function]}
            onKeyDown={[Function]}
            title="Expand/Collapse"
            type="button"
          >
            <svg
              aria-label="Expand/Collapse"
              className="bx--accordion__arrow"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="11,8 6,13 5.3,12.3 9.6,8 5.3,3.7 6,3"
              />
            </svg>
            <div
              className="bx--accordion__title"
            >
              Customize commit (advanced)
            </div>
          </button>
          <div
            className="bx--accordion__content"
          >
            <div
              className="bx--row margin-top-06"
            >
              <div
                className="bx--col"
              >
                <p>
                  Perform commit
                </p>
                <div
                  className="bx--form-item"
                >
                  <input
                    aria-label={null}
                    className="bx--toggle-input"
                    defaultChecked={true}
                    id="commitToggle"
                    onChange={[Function]}
                    onKeyUp={[Function]}
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle"
                    className="bx--toggle-input__label"
                    htmlFor="commitToggle"
                  >
                    <span
                      className="bx--toggle__switch"
                    >
                      <span
                        aria-hidden="true"
                        className="bx--toggle__text--off"
                      >
                        Off
                      </span>
                      <span
                        aria-hidden="true"
                        className="bx--toggle__text--on"
                      >
                        On
                      </span>
                    </span>
                  </label>
                </div>
              </div>
              <div
                className="bx--col"
              >
                <p>
                  Peers to endorse commit transactions
                </p>
                <div
                  className="bx--multi-select__wrapper bx--list-box__wrapper"
                >
                  <div
                    className="bx--multi-select bx--multi-select--selected bx--list-box"
                    id="carbon-multiselect-example"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                  >
                    <div
                      aria-controls={null}
                      aria-disabled={false}
                      aria-expanded={false}
                      aria-haspopup={true}
                      aria-labelledby="multiselect-label-1 multiselect-field-label-1"
                      aria-owns={null}
                      className="bx--list-box__field"
                      data-toggle={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                      tabIndex="0"
                      type="button"
                    >
                      <div
                        aria-label="Clear Selection"
                        className="bx--list-box__selection bx--tag--filter bx--list-box__selection--multi"
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        role="button"
                        tabIndex={0}
                        title="Clear all selected items"
                      >
                        2
                        <svg
                          aria-hidden={true}
                          focusable="false"
                          height={16}
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width={16}
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <polygon
                            points="24 9.4 22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4"
                          />
                        </svg>
                      </div>
                      <span
                        className="bx--list-box__label"
                        id="multiselect-field-label-1"
                      >
                        MultiSelect Label
                      </span>
                      <div
                        className="bx--list-box__menu-icon"
                      >
                        <svg
                          aria-label="Open menu"
                          focusable="false"
                          height={16}
                          name="chevron--down"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 16 16"
                          width={16}
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <polygon
                            points="8,11 3,6 3.7,5.3 8,9.6 12.3,5.3 13,6"
                          />
                          <title>
                            Open menu
                          </title>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="bx--row margin-top-07"
            >
              <div
                className="bx--col"
              >
                <div
                  className="bx--data-table-container"
                >
                  <table
                    className="bx--data-table bx--data-table--short bx--data-table--no-border"
                  >
                    <thead>
                      <tr>
                        <th
                          scope="col"
                        >
                          <span
                            className="bx--table-header-label"
                          >
                            Organization
                          </span>
                        </th>
                        <th
                          scope="col"
                        >
                          <span
                            className="bx--table-header-label"
                          >
                            Approval status
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      aria-live="polite"
                    >
                      <tr>
                        <td>
                          Hardcoded
                        </td>
                        <td>
                          Value
                        </td>
                      </tr>
                      <tr>
                        <td>
                          Needs to be
                        </td>
                        <td>
                          Changed
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            <div
              className="bx--row margin-top-07"
            >
              <div
                className="bx--col"
              >
                <p>
                  For an explanation of advanced scenarios, see the 
                  <a
                    className="bx--link"
                    href="#"
                  >
                    documentation
                  </a>
                  .
                </p>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>,
]
`;
